profiling/src/profile/mod.rs

```diff
@@ -81,6 +83,42 @@ struct Sample {
     local_root_span_id_label_offset: Option<usize>,
 }
 
+struct UpscalingRule {
+    values_offset: Vec<usize>,
+    label_name: i64,
+    label_value: i64,
+    upscaling_info: UpscalingInfo,
+}
+
+impl UpscalingRule {
+    pub fn compute_scale(&self, values: &[i64]) -> f64 {
+        match self.upscaling_info {
+            UpscalingInfo::Poisson {
+                x,
+                y,
+                sampling_distance,
+            } => {
+                if values[y] == 0 {
+                    return 0_f64;
+                }
+
+                let avg = (values[x] / values[y]) as f64;
+                1_f64 / (1_f64 - (-avg / sampling_distance as f64).exp())
+            }
+            UpscalingInfo::Proportional {
+                total_sampled,
+                total_real,
+            } => {
+                if total_sampled == 0 {
+                    return 0_f64;
+                }
+
+                total_real as f64 / total_sampled as f64
+            }
```

Comment:
Should this perhaps be precalculated when creating the internal `UpscalingRule`? Unlike the poisson approach, this is always going to be a constant (perhaps we don't even need to store the sampled/read internally -- even though I agree that it's nicer if the API caller specifies them separately).

Reply:
You are right, I was wondering if I can cache it when access the first time or like you said, we know it will result in a constant, and compute it when adding it.

Reply:
In the end, when going from FFI to Rust `pure` (do not know the term), I compute the scale once and for all.